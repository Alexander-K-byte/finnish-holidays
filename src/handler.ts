import { addDays, format } from "date-fns"

export interface Holiday {
    date: string;
    name: string;
    type: "public" | "observance"
}

const formatDate = (d: Date): string => format(d, "dd-MM-yyyy");

const getFixedHolidays = (year: number): Holiday[] => {
    const fixed = [
        { day: 1, month: 0, name: "New Year's Day" },
        { day: 6, month: 0, name: "Epiphany" },
        { day: 1, month: 4, name: "May Day" },
        { day: 4, month: 10, name: "All Saints' day" },
        { day: 6, month: 11, name: "Independence day" },
        { day: 25, month: 11, name: "Christmas Day" },
        { day: 26, month: 11, name: "Boxing Day" }
    ];

    return fixed.map(f => ({
        date: formatDate(new Date(year, f.month, f.day)),
        name: f.name,
        type: "public"
    }));
};

/**
 * computus algorithm for calculating Easter holidays
 */
const getEasterSunday = (year: number): Date => {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const month = Math.floor((h + l - 7 * m + 114) / 31) - 1; // zero-indexed
    const day = ((h + l - 7 * m + 114) % 31) + 1;
    return new Date(year, month, day);
};

const getGoodFriday = (year: number): Holiday => {
    const easter = getEasterSunday(year);
    const date = addDays(easter, - 2);
    return { date: formatDate(date), name: "Good Friday", type: "public" };
};

const getEasterMonday = (year: number): Holiday => {
    const easter = getEasterSunday(year);
    const date = addDays(easter, +1);
    return { date: formatDate(date), name: "Easter Monday", type: "public" };
};

/**
 * Ascension day is 40 days after easter Sunday
 */
const getAscensionDay = (year: number): Holiday => {
    const easter = getEasterSunday(year);
    const date = addDays(easter, +40);
    return { date: formatDate(date), name: "Ascension day", type: "public" };
};

/**
 * Pentecost is 10 days after Ascension day or 50 days after easter
 */
const getPentecost = (year: number): Holiday => {
    const easter = getEasterSunday(year);
    const date = addDays(easter, +50);
    return { date: formatDate(date), name: "Pentecost", type: "public" };
};

/**
 * Midsummer, first Saturday 20-26 June
 */
const getMidsummer = (year: number) => {
    const start = new Date(year, 5, 20);
    for (let i = 0; i <= 6; i++) {
        const a = addDays(start, i);
        if (a.getDay() === 6) {
            return { date: formatDate(a), name: "Midsummer", type: "public" }
        }
    }
};

export {
    formatDate,
    getFixedHolidays,
    getEasterSunday,
    getEasterMonday,
    getGoodFriday,
    getAscensionDay,
    getPentecost,
    getMidsummer
};