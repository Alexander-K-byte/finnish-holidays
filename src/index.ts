import {
    formatDate,
    getFixedHolidays,
    getEasterHolidays,
    getMidsummer,
    type Holiday
} from "./handler";

/**
 * Returns all Finnish public holidays for a given year.
 * Includes fixed holidays, Easter-related holidays, and Midsummer.
 * @param year Year to get holidays for
 */
const getHolidays = (year: number): Holiday[] => {
    return [
        ...getFixedHolidays(year),
        ...getEasterHolidays(year),
        getMidsummer(year)
    ];
};

export {
    type Holiday,
    getHolidays,
    getFixedHolidays,
    getEasterHolidays,
    getMidsummer,
    formatDate
};
