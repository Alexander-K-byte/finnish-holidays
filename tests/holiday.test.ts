import { describe, it, expect, beforeAll } from "vitest";
import { getHolidays, type Holiday } from "../src/index";
import { DateTime } from "luxon";

const testYears = [2024, 2025, 2026, 2032]; // include leap year

describe("Finnish Holidays Library", () => {
    testYears.forEach(year => {
        describe(`Year ${year}`, () => {

            let holidays: Holiday[];

            beforeAll(() => {
                holidays = getHolidays(year);
            });

            it("includes all fixed-date public holidays", () => {
                const fixedNames = [
                    "New Year's Day",
                    "Epiphany",
                    "May Day",
                    "All Saints' Day",
                    "Independence Day",
                    "Christmas Day",
                    "Boxing Day",
                ];
                fixedNames.forEach(name => {
                    const holiday = holidays.find(h => h.name === name);
                    expect(holiday).toBeDefined();
                    expect(holiday?.type).toBe("public");
                });
            });

            it("includes all Easter-related holidays", () => {
                const movingNames = ["Good Friday", "Easter Monday", "Ascension Day", "Pentecost"];
                movingNames.forEach(name => {
                    const holiday = holidays.find(h => h.name === name);
                    expect(holiday).toBeDefined();
                    expect(holiday?.type).toBe("public");
                });
            });

            it("includes Midsummer Day", () => {
                const midsummer = holidays.find(h => h.name === "Midsummer Day");
                expect(midsummer).toBeDefined();
                expect(midsummer?.type).toBe("public");

                // Ensure it's between June 20-26
                const midsummerDate = DateTime.fromFormat(midsummer!.date, "dd-MM-yyyy");
                expect(midsummerDate.month).toBe(6);
                expect(midsummerDate.day).toBeGreaterThanOrEqual(20);
                expect(midsummerDate.day).toBeLessThanOrEqual(26);
                expect(midsummerDate.weekday).toBe(6); // Saturday
            });

            it("aggregated getHolidays returns correct total count", () => {
                // 7 fixed + 4 Easter-related + 1 Midsummer
                const expectedCount = 7 + 4 + 1;
                expect(holidays.length).toBe(expectedCount);
            });

            it("matches snapshot for full holiday list", () => {
                expect(holidays).toMatchSnapshot();
            });

            it("has correctly formatted dates", () => {
                holidays.forEach(h => {
                    // Must match dd-MM-yyyy
                    expect(h.date).toMatch(/^\d{2}-\d{2}-\d{4}$/);
                });
            });

            it("returns consistent dates for fixed holidays across multiple years", () => {
                const independenceDay = holidays.find(h => h.name === "Independence Day");
                expect(independenceDay?.date.endsWith(`-${year}`)).toBe(true);
            });
        });
    });
});
