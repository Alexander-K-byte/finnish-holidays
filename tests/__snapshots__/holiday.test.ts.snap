// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Finnish Holidays Library > Year 2024 > matches snapshot for full holiday list 1`] = `
[
  {
    "date": "01-01-2024",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2024",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2024",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2024",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2024",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2024",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2024",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "29-03-2024",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "01-04-2024",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "09-05-2024",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "19-05-2024",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "22-06-2024",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays Library > Year 2025 > matches snapshot for full holiday list 1`] = `
[
  {
    "date": "01-01-2025",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2025",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2025",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2025",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2025",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2025",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2025",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "18-04-2025",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "21-04-2025",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "29-05-2025",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "08-06-2025",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "21-06-2025",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays Library > Year 2026 > matches snapshot for full holiday list 1`] = `
[
  {
    "date": "01-01-2026",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2026",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2026",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2026",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2026",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2026",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2026",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "03-04-2026",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "06-04-2026",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "14-05-2026",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "24-05-2026",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "20-06-2026",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays Library > Year 2032 > matches snapshot for full holiday list 1`] = `
[
  {
    "date": "01-01-2032",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2032",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2032",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2032",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2032",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2032",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2032",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "26-03-2032",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "29-03-2032",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "06-05-2032",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "16-05-2032",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "26-06-2032",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays for 2024 > should match the full holiday list snapshot 1`] = `
[
  {
    "date": "01-01-2024",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2024",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2024",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2024",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2024",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2024",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2024",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "29-03-2024",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "01-04-2024",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "09-05-2024",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "19-05-2024",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "22-06-2024",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays for 2025 > should match the full holiday list snapshot 1`] = `
[
  {
    "date": "01-01-2025",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2025",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2025",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2025",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2025",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2025",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2025",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "18-04-2025",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "21-04-2025",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "29-05-2025",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "08-06-2025",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "21-06-2025",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;

exports[`Finnish Holidays for 2026 > should match the full holiday list snapshot 1`] = `
[
  {
    "date": "01-01-2026",
    "name": "New Year's Day",
    "type": "public",
  },
  {
    "date": "06-01-2026",
    "name": "Epiphany",
    "type": "public",
  },
  {
    "date": "01-05-2026",
    "name": "May Day",
    "type": "public",
  },
  {
    "date": "04-11-2026",
    "name": "All Saints' Day",
    "type": "public",
  },
  {
    "date": "06-12-2026",
    "name": "Independence Day",
    "type": "public",
  },
  {
    "date": "25-12-2026",
    "name": "Christmas Day",
    "type": "public",
  },
  {
    "date": "26-12-2026",
    "name": "Boxing Day",
    "type": "public",
  },
  {
    "date": "03-04-2026",
    "name": "Good Friday",
    "type": "public",
  },
  {
    "date": "06-04-2026",
    "name": "Easter Monday",
    "type": "public",
  },
  {
    "date": "14-05-2026",
    "name": "Ascension Day",
    "type": "public",
  },
  {
    "date": "24-05-2026",
    "name": "Pentecost",
    "type": "public",
  },
  {
    "date": "20-06-2026",
    "name": "Midsummer Day",
    "type": "public",
  },
]
`;
